File: 프로그래머스/1/118666. 성격 유형 검사하기/성격 유형 검사하기.java
import java.util.Arrays;
import java.util.List;

class Solution {
    public String solution(String[] survey, int[] choices) {
        String[] grade1 = {"R", "C", "J", "A"};
        String[] grade2 = {"T", "F", "M", "N"};
        
        int[] score = new int[4];
        
        int[] dic = {0, -3, -2, -1, 0, 1, 2, 3}; // 0번 인덱스는 버림

        // 설문 데이터 처리
        for (int i = 0; i < survey.length; i++) {
            String s = survey[i];
            int choice = choices[i];
            
            // 현재 질문이 몇 번째 지표(R, C, J, A)인지 찾기
            int idx = -1;
            for (int j = 0; j < 4; j++) {
                if (s.contains(grade1[j])) {
                    idx = j;
                    break;
                }
            }

            // 질문의 순서가 뒤집혀 있는지 확인 (예: "TR")
            int currentScore = dic[choice];
            if (s.startsWith(grade2[idx])) { 
                // 첫 글자가 grade2(T, F, M, N)라면 점수 부호를 반대로
                currentScore *= -1;
            }
            
            score[idx] += currentScore;
        }

        StringBuilder answer = new StringBuilder();
        for (int i = 0; i < 4; i++) {
            // 점수가 0보다 작거나 같으면 grade1 (R, C, J, A)
            // 0보다 크면 grade2 (T, F, M, N)
            if (score[i] <= 0) {
                answer.append(grade1[i]);
            } else {
                answer.append(grade2[i]);
            }
        }

        return answer.toString();
    }
}

File: 프로그래머스/1/118666. 성격 유형 검사하기/성격 유형 검사하기.py
def solution(survey, choices):
    # 각 지표의 기준이 되는 글자
    standard = ["R", "C", "J", "A"]
    oppose = ["T", "F", "M", "N"]
    scores = [0, 0, 0, 0]
    
    for i, s in enumerate(survey):
        point = choices[i] - 4 
        
        # 질문의 왼쪽 글자가 기준 리스트(standard)의 어디에 있는지 확인
        if s[0] in standard:
            idx = standard.index(s[0])
            scores[idx] += point # 기준 글자가 왼쪽에 있으면 point 그대로 더함
        else:
            idx = oppose.index(s[0])
            scores[idx] -= point # 기준 글자가 오른쪽에 있으면 point를 빼줌
            
    answer = ""
    for i, s in enumerate(scores):
        if s <= 0: # 0보다 작거나 같으면 왼쪽 글자
            answer += standard[i]
        else:
            answer += oppose[i]
            
    return answer